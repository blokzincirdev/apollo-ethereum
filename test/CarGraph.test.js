const gql = require('graphql-tag')

const Car = artifacts.require('Car')
const CarReview = artifacts.require('CarReview')

const config = {
  contracts: {
    Car: {
      fields: {
        firstReview: 'CarReview',
        allReviews: 'CarReview',
        ReviewAdded: {
          type: 'CarReview',
          field: 'address',
        },
      },
    },
    CarReview: true,
  },
}

const QUERY_FIRST_REVIEW = gql`
  query($address: Address!) {
    Car(address: $address) {
      name
      firstReview {
        rating
        review
      }
    }
  }
`

const QUERY_ALL_REVIEWS = gql`
  query($address: Address) {
    Car(address: $address) {
      allReviews {
        rating
        review
      }
    }
  }
`

const QUERY_REVIEW_EVENTS = gql`
  query($address: Address) {
    Car(address: $address) {
      ReviewAdded {
        rating
        review
      }
    }
  }
`

contract('CarGraph', () => {
  let client

  before(async () => {
    client = await createClient(config, { Car, CarReview })
  })

  it('generates graphql schema', async () => {
    expect(client.source).to.matchSnapshot()
  })

  it('links address fields', async () => {
    const res = await client.execute(QUERY_FIRST_REVIEW, {
      address: Car.address,
    })
    expect(res.Car).to.deep.equal({
      name: 'Tesla',
      firstReview: {
        rating: 'Good',
        review: 'This car rocks',
      },
    })
  })

  it('links array of addresses', async () => {
    const res = await client.execute(QUERY_ALL_REVIEWS, {
      address: Car.address,
    })
    expect(res.Car.allReviews).to.deep.equal([
      {
        rating: 'Good',
        review: 'This car rocks',
      },
      {
        rating: 'Bad',
        review: 'This car does not rock',
      },
    ])
  })

  it('links addresses from event fields', async () => {
    const res = await client.execute(QUERY_REVIEW_EVENTS, {
      address: Car.address,
    })
    expect(res.Car.ReviewAdded).to.deep.equal([
      {
        rating: 'Good',
        review: 'This car rocks',
      },
      {
        rating: 'Bad',
        review: 'This car does not rock',
      },
    ])
  })
})
