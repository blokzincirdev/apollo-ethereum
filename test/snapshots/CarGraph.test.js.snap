// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract: CarGraph generates graphql schema 1`] = `
"directive @erc1820(interfaceName: String) on FIELD_DEFINITION

directive @mappingIndex(mapping: String) on FIELD_DEFINITION

directive @contract(field: String) on FIELD_DEFINITION

scalar Address

scalar BigNumber

type Car {
  _address: Address
  ReviewAdded: [CarReview] @contract(field: \\"address\\")
  firstReview: CarReview @contract
  name: String
  reviews(index: BigNumber!): Address
  allReviews: [CarReview] @contract
}

type CarMutative {
  addReview(address: Address): Boolean
}

type CarReview {
  _address: Address
  rating: CarReview_Rating
  review: String
}

enum CarReview_Rating {
  Bad
  Good
}

type Mutation {
  Car(address: Address!): CarMutative
}

type Query {
  Car(address: Address!): Car
  CarReview(address: Address!): CarReview
}
"
`;
